store/tikv/2pc.go

```diff
@@ -135,6 +146,29 @@ func newTwoPhaseCommitter(txn *tikvTxn, connID uint64) (*twoPhaseCommitter, erro
 	if len(keys) == 0 {
 		return nil, nil
 	}
+
+	for _, pair := range txn.assertions {
+		mutation, ok := mutations[string(pair.key)]
+		if !ok {
+			log.Error("ASSERTION FAIL!!! assertion exists but no mutation?", pair)
+			continue
+		}
+		// Only apply the first assertion!
+		// TODO: Find a way to avoid mutationEx.
```

Comment:
To avoid `mutationEx`, you can use a temp map to store keys that already have assertion. Or you can first stable sort + dedup `txn.assertions` before iterate it.

Reply:
Or we can stable sort, then dedup during iteration.

Reply:
mutationEx cause no performance burden:

```
type mutationEx struct {
    pb.Mutation
    asserted bool
}

var tmp mutationEx
mutation := &tmp.Mutation
```

However, use a temp map does have addition cost: it makes more allocation.
And sort has its cost too @disksing @coocood